library/std/src/panicking.rs

```diff
@@ -624,6 +698,83 @@ fn rust_panic_with_hook(
+/// abort or unwind.
+#[cfg(not(bootstrap))]
+fn rust_panic_error_with_hook(
+    error: &(dyn crate::error::Error + 'static),
+    location: &Location<'_>,
+) -> ! {
+    let (must_abort, panics) = panic_count::increase();
+
+    // If this is the third nested call (e.g., panics == 2, this is 0-indexed),
+    // the panic hook probably triggered the last panic, otherwise the
+    // double-panic check would have aborted the process. In this case abort the
+    // process real quickly as we don't want to try calling it again as it'll
+    // probably just panic again.
+    if must_abort || panics > 2 {
+        if panics > 2 {
+            // Don't try to print the message in this case
+            // - perhaps that is causing the recursive panics.
+            rtprintpanic!("thread panicked while processing panic. aborting.\n");
+        } else {
+            // Unfortunately, this does not print a backtrace, because creating
+            // a `Backtrace` will allocate, which we must to avoid here.
+            let panicinfo = PanicInfo::error_constructor(error, location);
+            rtprintpanic!("{}\npanicked after panic::always_abort(), aborting.\n", panicinfo);
+        }
+        intrinsics::abort()
+    }
+
+    unsafe {
+        let info = PanicInfo::error_constructor(error, location);
+        let _guard = HOOK_LOCK.read();
+        match HOOK {
+            // Some platforms (like wasm) know that printing to stderr won't ever actually
+            // print anything, and if that's the case we can skip the default
+            // hook. Since string formatting happens lazily when calling `payload`
+            // methods, this means we avoid formatting the string at all!
+            // (The panic runtime might still call `payload.take_box()` though and trigger
+            // formatting.)
+            Hook::Default if panic_output().is_none() => {}
+            Hook::Default => {
+                default_hook(&info);
+            }
+            Hook::Custom(ptr) => {
+                (*ptr)(&info);
+            }
+        };
+    }
+
+    if panics > 1 {
+        // If a thread panics while it's already unwinding then we
+        // have limited options. Currently our preference is to
+        // just abort. In the future we may consider resuming
+        // unwinding or otherwise exiting the thread cleanly.
+        rtprintpanic!("thread panicked while panicking. aborting.\n");
+        intrinsics::abort()
+    }
+
+    struct ErrorPanicPayload;
+
+    unsafe impl BoxMeUp for ErrorPanicPayload {
+        fn take_box(&mut self) -> *mut (dyn Any + Send) {
+            // ErrorPanicPayload is a zst so this box should not allocate
+            let data = Box::new(ErrorPanicPayload);
+            Box::into_raw(data)
+        }
+
+        fn get(&mut self) -> &(dyn Any + Send) {
+            &ErrorPanicPayload
+        }
+    }
+
+    rust_panic(&mut ErrorPanicPayload)
```

Comment:
Ideally it should be possible to panic with no payload at all. My intuition is that propagating the error with an unwinding panic to be caught and potentially reacted to is an anti-pattern, since converting a `dyn Error` to a panic semantically represents promoting a recoverable error to a non-recoverable error. I'm interested in hearing real world usecases of propagating and handling payloads via panics and am open to the possibility that I'm making the wrong call here, but my expectation is that if you need to react to an error in a panic it should never have been a panic in the first place.

For now we make a dummy payload that shouldn't need to allocate and which only indicates that the current panic started as a dyn Error.